Annually reset returns invoices

josAlba opened the pull request:
Update, to reset the subscription number every year.

<!-----------------------------------------------------------------------------
Thank you for contributing to the PrestaShop project! 

Please take the time to edit the "Answers" rows below with the necessary information.

Check out our contribution guidelines to find out how to complete it:
https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests
------------------------------------------------------------------------------>

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Change, to restart the return invoices annually. Since it currently takes the invoice number by the id of the autoincremental database. In addition to being able to give errors if any record is deleted or exchanged. The proposed change posts the invoices created within the same year, in this way the autoincremental id is not taken into account. It takes into account the order of the invoice in the year of the same.  
| Type?             | improvement
| Category?         |  BO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | Fixes #23080
| How to test?      | It can be verified by downloading the invoice pdf by subscription
| Possible impacts? | PDF invoices by subscription


<!-- Click the form's "Preview" button to make sure the table is functional in GitHub. Thank you! -->

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/prestashop/prestashop/23075)
<!-- Reviewable:end -->

josAlba commented:
> Hi, thanks for this contribution!
> 
> I found some issues with the Pull Request description:
> 
>     * The `category` should be one of these: `BO`, `CO`, `FO`, `IN`, `LO`, `ME`, `PM`, `TE` or `WS` ([Read explanation](https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/pull-requests/#type--category))
> 
> 
> Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
> 
> (Note: this is an automated message, but answering it will reach a real human)

This update affects when downloading the subscription pdf in the backoffice

| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop / 1.7.7.x
| Category?         |  BO

PierreRambaud commented:
> > Hi, thanks for this contribution!
> > I found some issues with the Pull Request description:
> > ```
> > * The `category` should be one of these: `BO`, `CO`, `FO`, `IN`, `LO`, `ME`, `PM`, `TE` or `WS` ([Read explanation](https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/pull-requests/#type--category))
> > ```
> > 
> > 
> > Would you mind having a look at it? This will help us understand how interesting your contribution is, thank you very much!
> > (Note: this is an automated message, but answering it will reach a real human)
> 
> This update affects when downloading the subscription pdf in the backoffice
> 
> Questions	Answers
> Branch?	develop / 1.7.7.x
> Category?	BO

Hi and thanks for your contribution.

Please fill in the description of your pull request (the first message).

Thanks

PierreRambaud reviewed (changes requested):
Please correctly fill the form and use PrestaShop coding standard :wink:

josAlba commented:
> Please correctly fill the form and use PrestaShop coding standard 😉

Hi, sorry I just updated it. Sorry if there are any grammatical errors, I use the translator.

PierreRambaud commented:
> > Please correctly fill the form and use PrestaShop coding standard wink
> 
> Hi, sorry I just updated it. Sorry if there are any grammatical errors, I use the translator.

You have to write code in English, otherwise, we won't be able to review it :wink:

josAlba commented:
> You have to write code in English, otherwise, we won't be able to review it 😉

I've been translating comments, etc. before uploading them.

> > > Please correctly fill the form and use PrestaShop coding standard wink
> > 
> > 
> > Hi, sorry I just updated it. Sorry if there are any grammatical errors, I use the translator.
> 
> You have to write code in English, otherwise, we won't be able to review it 😉



> > > Please correctly fill the form and use PrestaShop coding standard wink
> > 
> > 
> > Hi, sorry I just updated it. Sorry if there are any grammatical errors, I use the translator.
> 
> You have to write code in English, otherwise, we won't be able to review it 😉

I've been translating comments, etc. before uploading them.

PierreRambaud commented:
> I've been translating comments, etc. before uploading them.

Even the code has to be in English and follow our [coding standards](https://devdocs.prestashop.com/1.7/development/coding-standards/)

josAlba commented:
> > I've been translating comments, etc. before uploading them.
> 
> Even the code has to be in English and follow our [coding standards](https://devdocs.prestashop.com/1.7/development/coding-standards/)

I have rewritten the code in English, and modified the name of the variables.

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,32 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+     *
+     * @return int number for year
+     */
+    public function getNumberOrderSlip($id_order_slip)
```
```suggestion
    public function getNumberOrderSlip(int $id_order_slip): int
```

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,32 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
         $this->date = Tools::displayDate($this->order_slip->date_add);
         $prefix = Configuration::get('PS_CREDIT_SLIP_PREFIX', Context::getContext()->language->id);
-        $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, (int) $this->order_slip->id);
+        $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, (int) $this->getNumberOrderSlip($order_slip->id));
```
```suggestion
        $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, $this->getNumberOrderSlip((int) $order_slip->id));
```

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,32 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+    {
+        $temporary_order_slip = Db::getInstance()->executeS('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
+
+        $current_day = date('Y', strtotime($temporary_order_slip[0]['date_add']));
+        $current_year = $current_day . '-01-01 00:00:00';
```
```suggestion
        $current_year = date('Y-01-01 00:00:00', strtotime($temporary_order_slip[0]['date_add']));
```

Progi1984 reviewed (commented):
Some feedbacks. Thanks :pray:

josAlba commented:
> Some feedbacks. Thanks 🙏

Hi, thanks and updated with the suggested changes

josAlba commented:
> Please correctly fill the form and use PrestaShop coding standard 😉

Hi, I updated the form and made modifications to the code.

PierreRambaud reviewed (changes requested):
Hey, thanks for your update, but some things are missing in your pull request.
Following our [Contribution guidelines](https://devdocs.prestashop.com/1.7/contribute/contribution-guidelines/#pull-requests), you have to set a good title, something relevant, helping us to understand why you do that and why we have to review your change and of course in English :sweat_smile: . 
Then we will able to properly take time to review your pull request and make sure everything is fine :+1:

PierreRambaud commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,31 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+    {
+        $temporary_order_slip = Db::getInstance()->executeS('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
+
+        $current_year = date('Y-01-01 00:00:00', strtotime($temporary_order_slip[0]['date_add']));
```
You have to test if ` $temporary_order_slip[0]['date_add']` is available, same above

Please use camelCase instead of snake_case :)

josAlba replied:
Hi, and updated the code to perform the check and used camelCase. a greeting

Progi1984 commented:
The issue is waiting for PM.

marionf commented:
Hello @josAlba 

There is already an option in Order settings > Invoice > Invoice option to reset sequential invoice number at the beginning of the year

![capture d'écran_3014](https://user-images.githubusercontent.com/13449658/113310439-84aec600-9308-11eb-927a-77af9e37201b.png)

Did you know about this feature ?

josAlba commented:
> Hello @josAlba
> 
> There is already an option in Order settings > Invoice > Invoice option to reset sequential invoice number at the beginning of the year
> 
> ![capture d'écran_3014](https://user-images.githubusercontent.com/13449658/113310439-84aec600-9308-11eb-927a-77af9e37201b.png)
> 
> Did you know about this feature ?

Hello, for normal invoices yes but for credits there is no reset (orderSplit).
In invoices by subscription (orderSplit), in that case it always takes the reference by the id of the database

marionf commented:
@josAlba 

If I understand well, you want to be able to reset the credit slip number at the beginning of the year ?

josAlba commented:
Hello, I mean the "Delivery Slips", in which the identifier is retrieved from the id sent to be based on an annual numbering.

Sorry if I express myself wrong.

![imagen](https://user-images.githubusercontent.com/30494301/114040507-f7c5b880-9883-11eb-8358-88b5c200ab28.png)

In this way, the numbering corresponds to the season tickets in the same year.

![imagen](https://user-images.githubusercontent.com/30494301/114040827-4a9f7000-9884-11eb-82a4-858b36e27a27.png)

It is a problem that different users have had and it is also in the forum (in Spanish), where I contribute this solution that helped other users.

> @josAlba
> 
> If I understand well, you want to be able to reset the credit slip number at the beginning of the year ?
> 
> ![capture d'écran_3026](https://user-images.githubusercontent.com/13449658/114038030-cea42880-9881-11eb-92a7-5546d151cc45.png)

marionf commented:
Sorry but I still don't understand if you are talking about credit slip or delivery slip :thinking: 
You said you mean delivery slip but in your screenshot it's "factura por abono" 
"credit slip" = "factura por abono" 
"delivery slip" = "Albaranes de entrega"

josAlba commented:
hi, sorry if i meant "credit slip" 

> 
> 
> Sorry but I still don't understand if you are talking about credit slip or delivery slip 🤔
> You said you mean delivery slip but in your screenshot it's "factura por abono"
> "credit slip" = "factura por abono"
> "delivery slip" = "Albaranes de entrega"

marionf commented:
Hello @josAlba 

Thank you, it's more clear :+1:  I updated [the issue](https://github.com/PrestaShop/PrestaShop/issues/23080) title and description
Could you please update the title and description of this PR ? This will help for the code review and the tests.

I suggest to add a "Yes/no" toggle "Reset sequential credit slip number at the beginning of the year" in the BO > orders > credit slip > "Credit slip options" block to let the user choose if he/she wants to reset the number every year or not.

Like the option in the BO > orders > Invoices

![capture d'écran_3028](https://user-images.githubusercontent.com/13449658/114153674-df0ae080-991f-11eb-94e6-a7a6715d7e2e.png)

josAlba commented:
Hello, perfect I just added a check so that the user can make the decision. 

![imagen](https://user-images.githubusercontent.com/30494301/114174461-f1dddf00-9938-11eb-89c0-2f875c9a0327.png)

![imagen](https://user-images.githubusercontent.com/30494301/114174505-feface00-9938-11eb-8a5b-e1c4f580ef75.png)


> Hello @josAlba
> 
> Thank you, it's more clear 👍 I updated [the issue](https://github.com/PrestaShop/PrestaShop/issues/23080) title and description
> Could you please update the title and description of this PR ? This will help for the code review and the tests.
> 
> I suggest to add a "Yes/no" toggle "Reset sequential credit slip number at the beginning of the year" in the BO > orders > credit slip > "Credit slip options" block to let the user choose if he/she wants to reset the number every year or not.
> 
> Like the option in the BO > orders > Invoices
> 
> ![capture d'écran_3028](https://user-images.githubusercontent.com/13449658/114153674-df0ae080-991f-11eb-94e6-a7a6715d7e2e.png)

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+     */
+    public function getNumberOrderSlip(int $id_order_slip): int
+    {
+        $temporaryOrderSlip = Db::getInstance()->executeS('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
+
+        if (isset($temporaryOrderSlip[0]['date_add'])) {
+            $current_year = date('Y-01-01 00:00:00', strtotime($temporaryOrderSlip[0]['date_add']));
+
+            $numberOrderSlip = Db::getInstance()->executeS(
+                'SELECT count(*) as "n" FROM `' . _DB_PREFIX_ . 'order_slip` WHERE date_add > "' . $current_year . '" AND date_add < "' . $temporaryOrderSlip[0]['date_add'] . '"'
+            );
+
+            return intval($numberOrderSlip[0]['n']) + 1;
```
```suggestion
        $row = Db::getInstance()->executeS('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);

        if (isset($row[0]['date_add'])) {
            $current_year = date('Y-01-01 00:00:00', strtotime($row[0]['date_add']));

            $rowCount = Db::getInstance()->executeS(
                'SELECT count(*) as "n" FROM `' . _DB_PREFIX_ . 'order_slip` WHERE date_add > "' . $current_year . '" AND date_add < "' . $row[0]['date_add'] . '"'
            );

            return intval($rowCount[0]['n']) + 1;
```

Progi1984 reviewed (commented):
Some feedbacks

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CreditSlip/CreditSlipOptionsType.php:
```diff
@@ -43,22 +44,28 @@ final class CreditSlipOptionsType extends TranslatorAwareType
-        ]);
+            ])
+            ->add('slip_reset', SwitchType::class, [
+                'label' => $this->trans('Reset sequential invoice number at the beginning of the year', 'Admin.Orderscustomers.Feature'),
```
```suggestion
                'label' => $this->trans('Reset sequential credit slip number at the beginning of the year', 'Admin.Orderscustomers.Feature'),
```
I think you meant credit slip ?

josAlba replied:
Yes, I am sorry in Spain we call it something else. Therefore, it can be confusing when translating. I add the change thanks.

Julievrz commented on src/PrestaShopBundle/Form/Admin/Sell/Order/CreditSlip/CreditSlipOptionsType.php:
```diff
@@ -43,22 +44,28 @@ final class CreditSlipOptionsType extends TranslatorAwareType
+            ])
+            ->add('slip_reset', SwitchType::class, [
+                'label' => $this->trans('Reset sequential invoice number at the beginning of the year', 'Admin.Orderscustomers.Feature'),
+                'help' => $this->trans('Invoice number', 'Admin.Orderscustomers.Help'),
```
Hi, could you please remove this help text? :-)

josAlba replied:
Agree

josAlba reviewed (commented):
Update form

atomiix commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+     */
+    public function getNumberOrderSlip(int $id_order_slip): int
+    {
+        $row = Db::getInstance()->executeS('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
```
I think you can use `Db::getInstance()->getValue()` instead of `Db::getInstance()->executeS()`

josAlba replied:
OK. Perfect I make the change, you are right that way it is better thanks.

atomiix commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+        if (isset($row[0]['date_add'])) {
+            $current_year = date('Y-01-01 00:00:00', strtotime($row[0]['date_add']));
+
+            $rowCount = Db::getInstance()->executeS(
```
Same here

josAlba replied:
OK

atomiix reviewed (changes requested):
Some feedbacks, also, you should add the new configuration `PS_CREDIT_SLIP_RESET` in the file `install-dev/data/xml/configuration.xml`

josAlba reviewed (commented):
I have applied the changes

atomiix commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+     */
+    public function getNumberOrderSlip(int $id_order_slip): int
+    {
+        $date = Db::getInstance()->executeS('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
```
```suggestion
        $date = Db::getInstance()->getValue('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
```

atomiix reviewed (changes requested):
You forgot to change an `executeS` to a `getValue` ;) Can please resolve the conflict in the file `install-dev/data/xml/configuration.xml`?

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
         $prefix = Configuration::get('PS_CREDIT_SLIP_PREFIX', Context::getContext()->language->id);
-        $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, (int) $this->order_slip->id);
+
+        if (Configuration::get('PS_CREDIT_SLIP_RESET')) {
+            $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, $this->getNumberOrderSlip((int) $order_slip->id));
+        } else {
+            $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, (int) $order_slip->id);
+        }
```
```suggestion
        $this->title = sprintf(
            HTMLTemplateOrderSlip::l('%1$s%2$06d'),
            $prefix,
            Configuration::get('PS_CREDIT_SLIP_RESET') ? $this->getNumberOrderSlip((int) $order_slip->id) : (int) $order_slip->id
        );
```

Progi1984 reviewed (commented):
Don't forget to 1.7.9.0.sql the new configuration :)

josAlba commented:
> Don't forget to 1.7.9.0.sql the new configuration :)

Hello, Not having the file I have created it with the following structure, I don't know if that's what you mean.
<img width="714" alt="Captura de pantalla 2021-04-28 a las 12 04 54" src="https://user-images.githubusercontent.com/30494301/116386597-3a8f0680-a81a-11eb-8d47-2343146c2e5f.png">

atomiix commented on install-dev/data/xml/configuration.xml:
```diff
@@ -865,5 +865,14 @@ Country</value>
     <configuration id="PS_LOGS_EMAIL_RECEIVERS" name="PS_LOGS_EMAIL_RECEIVERS">
       <value>pub@prestashop.com</value>
     </configuration>
+    <configuration id="PS_SHOW_LABEL_OOS_LISTING_PAGES" name="PS_SHOW_LABEL_OOS_LISTING_PAGES">
+      <value>1</value>
+    </configuration>
+    <configuration id="ADDONS_API_MODULE_CHANNEL" name="ADDONS_API_MODULE_CHANNEL">
+      <value>stable</value>
+    </configuration>
```
That should not be in your PR 🤔

josAlba replied:
No, I added it to see if it solved the conflict since it indicated that my line would replace those two entries.

I removed them and left them as they were.

josAlba replied:
I already solved it, sorry

josAlba commented:
> 1.7.9.0.sql

I couldn't find the 1.7.9.0.sql file so I created it. I don't know if this is correct.

atomiix commented:
@josAlba you still have a conflict with the file `install-dev/data/xml/configuration.xml`. Also, if you can, it would be great to [squash your commits](https://levelup.gitconnected.com/how-to-squash-git-commits-9a095c1bc1fc)

josAlba commented:
> @josAlba you still have a conflict with the file `install-dev/data/xml/configuration.xml`. Also, if you can, it would be great to [squash your commits](https://levelup.gitconnected.com/how-to-squash-git-commits-9a095c1bc1fc)

I try to resolve the conflict but it won't let me. I understand that the fork will have a less recent file than the one that exists in the current branch to develop. But if I update my branch, it will ask me for a new pull request. 

<img width="805" alt="Captura de pantalla 2021-04-28 a las 12 32 16" src="https://user-images.githubusercontent.com/30494301/116391244-32859580-a81f-11eb-9293-5fa9afa9148b.png">

josAlba commented:
> to [squash your commits](https://levelup.gitconnected.com/how-to-squash-git-commits-9a095c1bc1fc)

Okay, I think I already have it. 

![imagen](https://user-images.githubusercontent.com/30494301/116394809-a75ace80-a823-11eb-8a15-7dd3af69a51c.png)

PierreRambaud commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+            return intval($counts) + 1;
+        }
+
+        return 1;
```
```suggestion
        return 0;
```

Otherwise, it makes no sense to return 1 when the id_order_slip is not found :thinking: 
The number should be 0 has there is no data.

josAlba replied:
It returns 1, because in the sql query it looks for a date less than the current invoice of the id_order_slip. So if there is only 1, the query will not return any items, therefore it is the first.

PierreRambaud replied:
So, the name of the method is not the good one :thinking:

josAlba replied:
It would be more representative if it were named: _getNumberOfYearOrderSlip_?

PierreRambaud commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+                'SELECT count(*) as "n" FROM `' . _DB_PREFIX_ . 'order_slip` WHERE date_add > "' . $current_year . '" AND date_add < "' . $date . '"'
+            );
+
+            return intval($counts) + 1;
```
```suggestion
            return intval($counts);
```

Same, I don't understand why you add 1 here :thinking:

PierreRambaud replied:
Or the name of the method is not the good one

josAlba replied:
As before, you have to add +1 to the retrieved value because it counts all except the id_order_slip.

PierreRambaud commented on install-dev/upgrade/sql/1.7.9.0.sql:
```diff
@@ -0,0 +1,5 @@
+SET NAMES 'utf8';
+
+INSERT INTO `PREFIX_configuration` (`name`, `value`, `date_add`, `date_upd`) VALUES
+('PS_CREDIT_SLIP_RESET', '0', NOW(), NOW());
```
```suggestion
('PS_CREDIT_SLIP_RESET', '0', NOW(), NOW());

```

PierreRambaud commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
     }
 
+    /**
+     * Returns the numbering of the subscription invoice with respect to the year.
```
I'm sorry but I don't get it :thinking:

PierreRambaud commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -67,11 +67,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+            $current_year = date('Y-01-01 00:00:00', strtotime($date));
+
+            $counts = Db::getInstance()->getValue(
+                'SELECT count(*) as "n" FROM `' . _DB_PREFIX_ . 'order_slip` WHERE date_add > "' . $current_year . '" AND date_add < "' . $date . '"'
```
I don't understand why you add `as "n"`, also I also don't understand why you are trying to find order_slip not related to the id_order_slip you have in parameter and not related to the order itself. It returns all order_slip in the database :thinking:

josAlba replied:
In this case, the 'as "n"' is no longer needed when using "getValue". I can leave it as "SELECT count (*) FROM` '", it was to give the column a value.

It would only retrieve the items that are between the selected ID and the calendar year to do a count.

![imagen](https://user-images.githubusercontent.com/30494301/116452183-b65e7280-a85d-11eb-8356-d856a93633be.png)

PierreRambaud replied:
You can remove the `"n"` which is useless

PierreRambaud reviewed (changes requested):
There are some mistake in your PR that we are not able to clearly understand :(

Progi1984 reviewed (changes requested):
Hello @josAlba, thanks for your contribution, would you mind to rebase your pull request? Thanks

Reference : https://anavarre.net/how-to-rebase-a-github-pull-request/

josAlba commented:
> Hello @josAlba, thanks for your contribution, would you mind to rebase your pull request? Thanks
> 
> Reference : https://anavarre.net/how-to-rebase-a-github-pull-request/

done and resolved the conflict in configuration.xml
Thank you

josAlba reviewed (commented):
Resolved

Julievrz reviewed (dismissed):
The wording is ✔️

atomiix commented:
@josAlba I'm sorry but it looks like your last commit is a merge commit 😅. You should rebase instead of merge

josAlba commented:
> @josAlba I'm sorry but it looks like your last commit is a merge commit 😅. You should rebase instead of merge

You're right, excuse me.
I have done the rebase
Can you confirm if everything is ok?
a greeting.

josAlba commented:
> @josAlba I'm sorry but it looks like your last commit is a merge commit 😅. You should rebase instead of merge

Hello, You're right I'm sorry.
I think the rebase is already done.

atomiix commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -68,11 +68,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+     *
+     * @return int number for year
+     */
+    public function getNumberOrderSlip(int $id_order_slip): int
```
```suggestion
    public function getOrderSlipIdForYear(int $id_order_slip): int
```

atomiix commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -68,11 +68,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+        $this->title = sprintf(
+            HTMLTemplateOrderSlip::l('%1$s%2$06d'),
+            $prefix,
+            Configuration::get('PS_CREDIT_SLIP_RESET') ? $this->getNumberOrderSlip((int) $order_slip->id) : (int) $order_slip->id
```
```suggestion
            Configuration::get('PS_CREDIT_SLIP_RESET') ? $this->getOrderSlipIdForYear((int) $order_slip->id) : (int) $order_slip->id
```

atomiix reviewed (commented):
Last feedbacks to try to make @PierreRambaud happy and we should be good 😄

PierreRambaud commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -68,11 +68,40 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+     */
+    public function getOrderSlipIdForYear(int $id_order_slip): int
+    {
+        $date = Db::getInstance()->getValue('SELECT date_add FROM `' . _DB_PREFIX_ . 'order_slip` WHERE id_order_slip=' . $id_order_slip);
```
So, if you select the wrong order slip, it returns 1. 
In this case, you should take the current year instead.

josAlba replied:
@PierreRambaud 
Hello,
I have changed so that if an id does not exist it returns 0, and to be able to filter by year. If the year is not added as a parameter, it will return the number corresponding to its own year.
¿Would this solution work?

PierreRambaud reviewed (changes requested):
Hi @josAlba 
Are you able to continue your PR about phpdoc, fix the CI and also the code.
About the $year in the parameter, what it's not used directly in the request, what's the aim of this one, you have to correctly explain it in order to be sure everyone have the same knowledge.

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -70,11 +76,44 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
         $prefix = Configuration::get('PS_CREDIT_SLIP_PREFIX', Context::getContext()->language->id);
-        $this->title = sprintf(HTMLTemplateOrderSlip::l('%1$s%2$06d'), $prefix, (int) $this->order_slip->id);
+
+        $OrderSlipId = (int) $order_slip->id;
```
```suggestion
        $orderSlipId = (int) $order_slip->id;
```

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -70,11 +76,44 @@ public function __construct(OrderSlip $order_slip, Smarty $smarty)
+        $this->title = sprintf(
+            HTMLTemplateOrderSlip::l('%1$s%2$06d'),
+            $prefix,
+            Configuration::get('PS_CREDIT_SLIP_RESET') ? $this->getOrderSlipNumberForYearByIdOrderSlip($OrderSlipId) : $OrderSlipId
```
```suggestion
            Configuration::get('PS_CREDIT_SLIP_RESET') ? $this->getOrderSlipNumberForYearByIdOrderSlip($orderSlipId) : $orderSlipId
```

Progi1984 commented on classes/pdf/HTMLTemplateOrderSlip.php:
```diff
@@ -159,18 +202,20 @@ public function getContent()
+                'order_slip' => $this->order_slip,
+                'order_details' => $this->order->products,
+                'cart_rules' => $this->order_slip->order_slip_type == 1 ? $this->order->getCartRules() : false,
+                'amount_choosen' => $this->order_slip->order_slip_type == 2 ? true : false,
```
```suggestion
                'amount_choosen' => $this->order_slip->order_slip_type == 2,
```

Progi1984 commented on install-dev/upgrade/sql/1.7.9.0.sql:
```diff
@@ -13,3 +13,9 @@ INSERT IGNORE INTO `PREFIX_hook` (`id_hook`, `name`, `title`, `description`, `po
+SET SESSION sql_mode='';
+SET NAMES 'utf8';
+
+INSERT INTO `PREFIX_configuration` (`name`, `value`, `date_add`, `date_upd`) VALUES
```
@josAlba Could you move this migration query to the module autoupgrade, please ?

Progi1984 reviewed (commented):
Some feedback

Commit:
Update credit slip options configuration

Also improve configuration.xml file

khouloudbelguith commented:
Hi @josAlba,

I have an issue following these steps:
1. Fresh install of this PR
2. Go to BO > Orders > Orders page
3. Change any Order status to shipped
4. Click on Partial refund > set quantity of a product and click on Partial refund 
![image](https://user-images.githubusercontent.com/16067358/141781482-cb4c4357-ce2d-4536-adf8-cce867d1df78.png)
5. A credit slip is created in the Documents tab 
![image](https://user-images.githubusercontent.com/16067358/141781852-efb869ab-8eee-44fe-a597-bc30fc68ac25.png)
6. Click to download  and see the credit slip
7. See error => empty pdf 
![image](https://user-images.githubusercontent.com/16067358/141781993-f81e1e4a-f402-4b09-ae2c-cfe7cdca2884.png)

Thanks!

PierreRambaud commented:
> Hi @josAlba,
> 
> I have an issue following these steps:
> 
> 1. Fresh install of this PR
> 2. Go to BO > Orders > Orders page
> 3. Change any Order status to shipped
> 4. Click on Partial refund > set quantity of a product and click on Partial refund
>    ![image](https://user-images.githubusercontent.com/16067358/141781482-cb4c4357-ce2d-4536-adf8-cce867d1df78.png)
> 5. A credit slip is created in the Documents tab
>    ![image](https://user-images.githubusercontent.com/16067358/141781852-efb869ab-8eee-44fe-a597-bc30fc68ac25.png)
> 6. Click to download  and see the credit slip
> 7. See error => empty pdf
>    ![image](https://user-images.githubusercontent.com/16067358/141781993-f81e1e4a-f402-4b09-ae2c-cfe7cdca2884.png)
> 
> Thanks!

I rebase the PR, but I'm not able to reproduce the issue, I'm trying to investigate what's wrong.
Did you do some additional steps?

PierreRambaud commented:
> Hi @PierreRambaud,
> 
> I confirm that I cannot reproduce the issue again.
> 
>  Untitled_.Jan.24.2022.7_41.PM.mp4 
> Thanks!

Since I didn't change anything instead of the rebase, maybe you can continue the QA validation?

khouloudbelguith reviewed (changes requested):
Hi @josAlba, @PierreRambaud,

I have an issue following these steps:
1. Install this PR with English language & France country
2. Create some orders & generate some credit slips (Year = 2022)
2. Go to the BO > Orders > Credit Slips page
3. Enable `Reset sequential credit slip number at the beginning of the year` and save 
![image](https://user-images.githubusercontent.com/16067358/153152564-ae1ba051-0fc9-4cbc-b034-f7cf1e72d166.png)
4. Change the time of your system to `31/12/2022 23:58:00`
5. Wait 2 minutes (System time = 01/01/2023) & create an order & generate a credit slip
6. Go to BO > Orders > View last order created > Documents Tab
7. See error: The number of the Credit slip is not reset 
![image](https://user-images.githubusercontent.com/16067358/153152917-025d1dc6-48ce-4ad9-a5fd-be72b28351b1.png)
8. Download the Credit Slip & view the pdf > The text of the credit's number in the pdf => ok 
![image](https://user-images.githubusercontent.com/16067358/153153214-b478e3ed-afcc-41ad-a3b8-17ed17aa077c.png)
9. Go to FO > Sign in with your customer & Go to  Your account >  Credit Slips
10. See error > The credit slips number is not reset 
![image](https://user-images.githubusercontent.com/16067358/153153353-5e72c9e9-067d-4879-9b31-87da723f844a.png)


I checked the same behavior with Invoice when enabling 
![image](https://user-images.githubusercontent.com/16067358/153153574-e0bcaf54-e012-45b9-bab9-127fb0c28f2b.png)
But the number => ok
![image](https://user-images.githubusercontent.com/16067358/153153784-d8ff7af7-26ff-4404-9353-78a9fee873b1.png)

Thanks!

Progi1984 commented:
Ping @josAlba for reminder

matks commented:
Hello,

It looks like this PR is now stale. I close it as our rules state it "Pull Requests may be closed after 30 days of inactivity following a request for modifications." (https://www.prestashop-project.org/maintainers-guide/reviewing-pull-requests/)

Thank you
